packages/ra-core/src/form/validate.ts

```diff
@@ -8,42 +7,47 @@ const EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"
 const isEmpty = (value: any) =>
     typeof value === 'undefined' || value === null || value === '';
 
+export interface ValidationErrorMessageWithArgs {
+    message: string;
+    args: {
+        [key: string]: ValidationErrorMessageWithArgs | any;
+    };
+}
+
+export type ValidationErrorMessage = string | ValidationErrorMessageWithArgs;
+
 export type Validator = (
     value: any,
     values: any,
     props: any
-) => string | null | undefined;
+) => ValidationErrorMessage | null | undefined;
 
 interface MessageFuncParams {
     args: any;
     value: any;
     values: any;
-    translate: Translate;
-    [key: string]: any;
 }
 
-type MessageFunc = (params: MessageFuncParams) => string;
+type MessageFunc = (params: MessageFuncParams) => ValidationErrorMessage;
 
 const getMessage = (
     message: string | MessageFunc,
     messageArgs: any,
     value: any,
-    values: any,
-    props: {
-        translate: Translate;
-    }
+    values: any
 ) =>
     typeof message === 'function'
         ? message({
               args: messageArgs,
               value,
               values,
-              ...props,
           })
-        : props.translate(message, {
-              _: message,
-              ...messageArgs,
-          });
+        : messageArgs
+        ? {
+              message,
+              args: messageArgs,
```

Comment:
you should pass the `message` as default translation (using the `_` arg) or users will be flooded by missing translation messages.

Reply:
Hu ? This is the job of the component which displays the error and actually, it does it